editor/store/test/selectors.js

```diff
@@ -2616,224 +2639,335 @@ describe( 'selectors', () => {
+			};
+			expect( canInsertBlockType( state, 'core/test-block-c', 'block1' ) ).toBe( false );
+		} );
+
+		it( 'should allow blocks that restrict parent to be inserted into an allowed parent', () => {
+			const state = {
+				editor: {
+					present: {
+						blocksByUID: {
+							block1: { name: 'core/test-block-b' },
 						},
-						edits: {},
 					},
 				},
-				currentPost: {},
-				sharedBlocks: {
-					data: {},
+				blockListSettings: {},
+				settings: {},
+			};
+			expect( canInsertBlockType( state, 'core/test-block-c', 'block1' ) ).toBe( true );
+		} );
+
+		it( 'should deny restricted blocks from being inserted into a block that restricts allowedBlocks', () => {
+			const state = {
+				editor: {
+					present: {
+						blocksByUID: {
+							block1: { name: 'core/test-block-a' },
+						},
+					},
 				},
+				blockListSettings: {
+					block1: {
+						supportedBlocks: [ 'core/test-block-c' ],
+					},
+				},
+				settings: {},
 			};
+			expect( canInsertBlockType( state, 'core/test-block-b', 'block1' ) ).toBe( false );
+		} );
 
-			const items = getInserterItems( state, [ 'core/test-block' ] );
-			expect( items[ 0 ].isDisabled ).toBe( true );
+		it( 'should allow allowed blocks to be inserted into a block that restricts allowedBlocks', () => {
+			const state = {
+				editor: {
+					present: {
+						blocksByUID: {
+							block1: { name: 'core/test-block-a' },
+						},
+					},
+				},
+				blockListSettings: {
+					block1: {
+						supportedBlocks: [ 'core/test-block-b' ],
+					},
+				},
+				settings: {},
+			};
+			expect( canInsertBlockType( state, 'core/test-block-b', 'block1' ) ).toBe( true );
 		} );
 
-		it( 'should properly list shared blocks', () => {
+		it( 'should prioritise parent over allowedBlocks', () => {
 			const state = {
 				editor: {
 					present: {
 						blocksByUID: {
-							carrot: { name: 'core/test-block' },
+							block1: { name: 'core/test-block-b' },
+						},
+					},
+				},
+				blockListSettings: {
+					block1: {
+						supportedBlocks: [],
+					},
+				},
+				settings: {},
+			};
+			expect( canInsertBlockType( state, 'core/test-block-c', 'block1' ) ).toBe( true );
+		} );
+	} );
+
+	describe( 'getInserterItems', () => {
+		it( 'should properly list block type and shared block items', () => {
+			const state = {
+				editor: {
+					present: {
+						blocksByUID: {
+							block1: { name: 'core/test-block-a' },
 						},
 						blockOrder: {},
 						edits: {},
 					},
 				},
-				currentPost: {},
 				sharedBlocks: {
 					data: {
-						123: { uid: 'carrot', title: 'My shared block' },
+						1: { uid: 'block1', title: 'Shared Block 1' },
 					},
 				},
+				currentPost: {},
+				preferences: {
+					insertUsage: {},
+				},
+				blockListSettings: {},
+				settings: {},
 			};
-
-			expect( getInserterItems( state, [ 'core/block' ] ) ).toEqual( [
+			expect( getInserterItems( state ) ).toEqual( [
 				{
-					id: 'core/block/123',
+					id: 'core/test-block-b',
```

Comment:
When looking only at this test it is unclear why Block B is listed before Block A. `utility` value inherited from the `common` category does the trick. Maybe Block B should be removed from this test and included in new test which tests only order of the items based on `utility` and `frecency`?

Reply:
👍 changed in 470d32e4b